Removing creation of predefined networks on swarm manager

abhinandanpb opened the pull request:
From 17.06 we started supporting service creation on host and bridge
predefined networks in addition to user defined local scope networks.
Since the task allocation involves creation of networks in the swarm
manager a network id is generated on the manager to create predefined networks.
But this network id is not visible with the api to fetch networks. Since
the task template which is part of a servicespec holds this network id
its hard to correlate the network id to network name by the api consumer.
So for predefined networks a predefined network id would make it easier
to correlate and update the service spec.
Corresponding moby PR: moby/moby#34278

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "hostnet" git@github.com:abhinandanpb/swarmkit.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

aaronlehmann commented:
swarmkit always uses pseudorandom object IDs with a consistent format. The IDs aren't meant for naming, so names should go in the `Annotations.Name` field instead. I don't think we should make this change.

sanimej commented:
@abhinandanpb @aaronlehmann Can we set `NetworkAttachmentConfig.Target` to the network name for predefined networks ? That looks like the natural thing to do because the network id in this case is not relevant or required. It should also address the UCP issue. Please see this discussion https://github.com/docker/orca/issues/7928#issuecomment-318091854

aaronlehmann commented:
If the impact on backward compatibility isn't too bad, I'm okay with it. I know there are currently some places in the network allocator that assume this field contains a network ID. I'm not sure if there are any places in the engine that assume this.

abhinandanpb commented:
@aluzzardi @aaronlehmann @mavenugo @sanime  
I have modified the changes. Hopefully with this we have arrived at a proper fix.
1) I have removed predefined networks from swarm manager
2) The service spec will have the target as ID,host,bridge
3) Allocator will skip the logic for predefined networks.
PTAL

aaronlehmann commented on manager/allocator/cnmallocator/networkallocator.go:
```diff
@@ -930,10 +930,29 @@ func initializeDrivers(reg *drvregistry.DrvRegistry) error {
+	return filteredNetworks
+}
+
+func taskNetworks(t *api.Task) []*api.NetworkAttachment {
```
I'd prefer to just add:

```go
		if IsPredefinedNetwork(n.Network.ID) {
			continue
		}
```

to loops over `t.Networks` which should skip predefined networks, including `releaseEndpoints`. That seems a little more explicit to me.

Otherwise, we should rename this so it's clear that it's only returning non-predefined networks.

aaronlehmann commented on manager/allocator/cnmallocator/networkallocator.go:
```diff
@@ -930,10 +930,29 @@ func initializeDrivers(reg *drvregistry.DrvRegistry) error {
+		if IsPredefinedNetwork(n.Target) {
+			continue
+		}
+		filteredNetworks = append(filteredNetworks, n)
```
As with `taskNetworks`, let's either change callers to explicitly skip predefined networks (there are only two callers), or rename this function to make it clear it's only returning a subset of the networks.

aaronlehmann commented:
Overall LGTM, just some minor comments on how we filter out predefined networks.

cc @ijc

aaronlehmann commented on manager/allocator/cnmallocator/drivers_network_linux.go:
```diff
@@ -26,3 +26,13 @@ func PredefinedNetworks() []networkallocator.PredefinedNetworkData {
 }
+
+// IsPredefinedNetwork checks if the network is a host/bridge network
+func IsPredefinedNetwork(target string) bool {
```
One more thought: this method isn't really necessary because callers can just call  `PredefinedNetworks` and check against that list (with the same few lines of code that exist here). Let's remove this in favor of `PredefinedNetworks` to avoid interface bloat.

abhinandanpb commented:
Closing the Pr in favor of moby/moby#34302
